### DISCUSSION TITLE
Override core shortcode in template

### DISCUSSION THREAD

**Discussion by rica-carv** [e107inc/e107#5284] [Q&A Themes]
How can i override a core template (present in e_shortcode) inside the theme_shortcode file?

I'm trying to override the PM_NAV shortcode, since it has some hardcoded HTML inside it, and i need to change that HTML in my theme....

Cretaing a function sc_pm_nav inside the theme_shortcode is not working for now....

**Comment by Jimmi08** [comment]
This was solved during time...
In theme_shortcodes there is property $override, it has to be true.

**Reply by rica-carv** [reply]
@Jimmi08 It's not working also, i tryed cleaning cache, restart server, and still nothing, with `var $override = true;` (as it is in landingzero theme_shortcodes) or `public $override = true;`

**Reply by rica-carv** [reply]
Just figure it out. If a shortcode is NOT in the e_shortcode file of a plugin, it's overriden, but if it is in the e_shortcode file of a plugin, it won't be overriden. 

For instance, i can't override the pm_nav shortcode from the pm plugin, because it is in the e_shortcode file.... but i can override the sign_in_pm_nav shortcode, because it is in the signin_shortcodes file from the signin plugin.... weird....

**Reply by Vodhin** [reply]
> For instance, i can't override the pm_nav shortcode from the pm plugin, because it is in the e_shortcode file.... but i can override the sign_in_pm_nav shortcode, because it is in the signin_shortcodes file from the signin plugin.... 

If this is for your own Theme, then just make your own custom PM Template file and include it with your Theme. Then you should be able to use your own sc_my_pm_nav() shortcode and reference it `{MY_PM_NAV}` in the template. I think that works.

**Reply by rica-carv** [reply]
@Vodhin I know i can totally rewrite a shortcode inside the theme, but if i only want to change a single line of code, it's like trying to get a shootgun to kill a mosquito...
If i go that way, i'll end up with a copy of all the e107 code inside my theme....

I don't want to go that way, i want to rewrite the minimum code i really need, and take use of what's already on the core.

Further, there's a lot of shortcodes that are not DIRECTLY available, for instance {NEWS_CATEGORY_COUNT_NEWS}.
I ended up mocking the system and manage to use it, since it's not part of shortcode file, but a instead of a class...
Even there's a lot of news shortcodes that doesn't work at theme level, and that's really annoying....

**Comment by Jimmi08** [comment]
@rica-carv 
I just replaced core shortcode from e_shortcode.php from siteinfo plugin and it works

```
class theme_shortcodes extends e_shortcode
{
    // public $override = true;
    var $override = true;

    function sc_sitelogo() {
        return "xxx";
    }
}
```

So it depends on something else.

**Reply by rica-carv** [reply]
@Jimmi08 Ok, you're right, i tested, and it works, but it works only if you use it in theme layout. Try to use it inside a custom template inside the theme... It won't work, the template for that plugin will use the plugin shortcodes and not the overriden ones in the theme_shortcode file.....

It's weird, but it's how it is now...

**Reply by Jimmi08** [reply]
The problem is - we (I) don't know how this should work - what is intention or rules.   So I don't know what is correct or what is not.  I have been asking for clear answers for years...   

There are:
- magic 
- global 
- plugin shortcodes (direct and batch)

All should be overridden in theme shortcodes.  
**magic shortcodes** are special shortcodes that should be used only in layouts and they are overridden in theme shortcodes. Confirmed. 
 
**Global shortcodes** -  I put {SITELOGO) into navigation_template and it still works. So it depends on the type of template, maybe? (core vs plugin?)  

**Plugin  shortcodes** - they should never use HTML code, only data, all HTML should go to templates...   so there should not be a reason to override them.  
 
I am almost sure that it is related to how shortcode and template is called in the plugin or in core - there are many parameters - like merged, replaced, only core...  with batch functionality - override, add, replace...   Almost every plugin is doing this different way, look at featurebox.  

Long ago I tried to extend download shortcodes, it is OOP, it should be easy...  ha, not at all.  

So now I ended with splitting news and members related functionality from core and replaced it with plugin. So it is hard for me to test anything nowadays.

**Reply by Vodhin** [reply]
`Plugin shortcodes - they should never use HTML code, only data…`

The whole point of Shortcodes is to generate standardized HTML.  It is there so you can do things like add an edit icon/link if the user has permissions to do that, or maybe you need an element with an ID so JavaScript can identify it, or data tags for AJAX calls. 

The difference here is that the HTML code generated by a short code should NOT contain classNames or style declarations to alter the appearance of an element unless absolutely needed, eg. setting the opacity of a “Like” icon based on wether or not a user liked something. You may also need special classNames to let your JavaScript to loop through groups of elements. 

So, yes. Shortcodes are meant to deliver HTML elements, to be wrapped by the HTML of a Template’s code, making it - in theory - easier to change the look of a website by styling the Template’s HTML.

**Reply by rica-carv** [reply]
@Vodhin Sorry, but i disagree. Shortcodes should deal with logic only, not have embebbed HTML. Rigth now there's a lot of shortcodes that have a lot of HTML embbebed, and it's only way to customize it is to create a whole new shortcodee for you to use in you theme. But if i can't override that shortcode in the theme, i have to recreate a whole new shortcode for doing that customization. It's not that hard do have shortcodes to output only logic, and leave ALL the HTML in the templates. I'm doing that all the time in my plugins (althought i know it won't be used by anyone else but me).
So on this i agree with @Jimmi08. Right now i'm ending up stuffing new shortcodes in my theme, when with a simple template customization they should work, but no.
Just see news shortcodes with HTML embebbed, Signin shortcodes idem, etc. The list is endless....
And please, don't tell me that is for retro compatibility, that can be donne by templating all that embebed HTML.
Further more, there's a lot of old plugins and theme that are already not working, so it should be mandatory to upgrade plugins and themes to new standards...
Windows 11 don't support anymore 32 bit computers, got my point?
